### PR TITLE
Update DtaScCreationFailedFixJob to address Appeals

### PR DESCRIPTION
**Pull request by AdamShawBAH**
Resolves [Jira Issue Title](https://jira.devops.va.gov/browse/JIRA-12345)

# Description
Please explain the changes you made here.

## Acceptance Criteria
- [ ] Code compiles correctly

## Testing Plan

1. Go to [Jira Issue/Test Plan Link](https://jira.devops.va.gov/browse/JIRA-12345) or list them below

- [ ] For feature branches merging into master: Was this deployed to UAT?

# Frontend
## User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 BEFORE|AFTER
 ---|---

## Storybook Story
*For Frontend (Presentation) Components*
* [ ] Add a [Storybook](https://github.com/department-of-veterans-affairs/caseflow/wiki/Documenting-React-Components-with-Storybook) file alongside the component file (e.g. create `MyComponent.stories.js` alongside `MyComponent.jsx`)
* [ ] Give it a title that reflects the component's location within the overall Caseflow hierarchy
* [ ] Write a separate story (within the same file) for each discrete variation of the component

# Backend
## Database Changes
*Only for Schema Changes*

* [ ] Add typical timestamps (`created_at`, `updated_at`) for new tables
* [ ] Update column comments; include a "PII" prefix to indicate definite or potential [PII data content](https://github.com/department-of-veterans-affairs/appeals-team/blob/master/caseflow-team/0-how-we-work/pii-handbook.md#what-is-pii)
* [ ] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`) (see [Writing DB migrations](https://github.com/department-of-veterans-affairs/caseflow/wiki/Writing-DB-migrations))
* [ ] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [ ] Perform query profiling (eyeball Rails log, check bullet and fasterer output)
* [ ] For queries using raw sql was an explain plan run by System Team
* [ ] Add appropriate indexes (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [ ] Run `make check-fks`; add any missing foreign keys or add to `config/initializers/immigrant.rb` (see [Record associations and Foreign Keys](https://github.com/department-of-veterans-affairs/caseflow/wiki/Record-associations-and-Foreign-Keys))
* [ ] Add `belongs_to` for associations to enable the [schema diagrams](https://department-of-veterans-affairs.github.io/caseflow/task_trees/schema/schema_diagrams) to be automatically updated
* [ ] Document any non-obvious semantics or logic useful for interpreting database data at [Caseflow Data Model and Dictionary](https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Data-Model-and-Dictionary)

## Integrations: Adding endpoints for external APIs
* [ ] Check that Caseflow's external API code for the endpoint matches the code in the relevant integration repo
  * [ ] Request: Service name, method name, input field names
  * [ ] Response: Check expected data structure
  * [ ] Check that calls are wrapped in MetricService record block
* [ ] Check that all configuration is coming from ENV variables
  * [ ] Listed all new ENV variables in description
  * [ ] Worked with or notified System Team that new ENV variables need to be set
* [ ] Update Fakes
* [ ] For feature branches: Was this tested in Caseflow UAT

# Best practices
## Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [ ] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Monitoring
- [ ] Are performance metrics (e.g., response time, throughput) being tracked?
- [ ] Are key application components monitored (e.g., database, cache, queues)?
- [ ] Is there a system in place for setting up alerts based on performance thresholds?

### Logging
- [ ] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 1]
```diff
@@ -1,36 +1,128 @@
...
 class DtaScCreationFailedFixJob < CaseflowJob
-  ERROR_TEXT = "DTA SC Creation Failed"
+  def initialize
+    @stuck_job_report_service = StuckJobReportService.new
+    @start_time = nil
+    @end_time = nil
+    super
+  end
+
+  def error_text
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:11]: why is this a function and not a constant

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 2]
```diff
@@ -1,36 +1,128 @@
...
+    "DTA SC creation failed"
+  end
 
   # :reek:FeatureEnvy
   def perform
-    stuck_job_report_service = StuckJobReportService.new
-    return if hlrs_with_errors.blank?
+    start_time
+
+    return if records_with_errors.blank?
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:19]: if this returns then you will not log the time? will that look right in the logs?

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 3]
```diff
@@ -1,36 +1,128 @@
...
+      "10"
+    else
+      fail ArgumentError, "Unsupported claimant type: #{type}"
+      # Or add other Claimant types if necessary
+    end
+  end
+
+  # :reek:FeatureEnvy
+  def process_records(record)
+    ActiveRecord::Base.transaction do
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:92]: why is this needed?

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 4]
```diff
@@ -1,36 +1,128 @@
...
+
+    @stuck_job_report_service.append_record_count(records_with_errors.count, error_text)
+
+    loop_through_and_call_process_records
+
+    @stuck_job_report_service.append_record_count(records_with_errors.count, error_text)
+    @stuck_job_report_service.write_log_report(error_text)
+
+    end_time
+    log_processing_time
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:29]: I don't see any high level error handling

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 5]
```diff
@@ -1,36 +1,128 @@
...
+  end
+
+  def claimant_type_to_payee_code(type)
+    case type
+    when "VeteranClaimant"
+      "00"
+    when "DependentClaimant"
+      "10"
+    else
+      fail ArgumentError, "Unsupported claimant type: #{type}"
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:85]: this will raise an error and will break your loop, is that your intention?

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 6]
```diff
@@ -1,36 +1,128 @@
...
+    @stuck_job_report_service.append_record_count(records_with_errors.count, error_text)
+    @stuck_job_report_service.write_log_report(error_text)
+
+    end_time
+    log_processing_time
+  end
+
+  def loop_through_and_call_process_records
+    process_records_for_hlrs_with_errors if hlrs_with_errors.present?
+    process_records_for_appeals_with_errors if appeals_with_errors.present?
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:34]: you are raising an error deeper in the call stack from this point, yet nothing is being rescued. This will generate an applicaion level error and logged in the application log but your error should really be a managed exception.

File: app/jobs/dta_sc_creation_failed_fix_job.rb (hunk 1 above)
Comment (AdamShawBAH) [reply app/jobs/dta_sc_creation_failed_fix_job.rb:11]: This was how it was updated for the stuck job scheduler, and for the interface. Could potentially go back to constants and put that in the shared_examples if you prefer

File: app/jobs/dta_sc_creation_failed_fix_job.rb (hunk 2 above)
Comment (AdamShawBAH) [reply app/jobs/dta_sc_creation_failed_fix_job.rb:19]: doesnt this make sense considering we would not want to run the job if there were no errors? therefore there shouldnt be logging of it anyway. in the reporter its just resolves to 0 if there is no time

File: app/jobs/dta_sc_creation_failed_fix_job.rb (hunk 4 above)
Comment (AdamShawBAH) [reply app/jobs/dta_sc_creation_failed_fix_job.rb:29]: what error handling are you looking for here? there is error handling that occurs within the process_records function which is the main operation. Is there some where else you want the error handling? do you mean to wrap loop_through_and_call_process_records?

File: app/jobs/dta_sc_creation_failed_fix_job.rb (hunk 6 above)
Comment (AdamShawBAH) [reply app/jobs/dta_sc_creation_failed_fix_job.rb:34]: i will wrap this in a begin rescue to catch and handle exceptions

File: app/jobs/dta_sc_creation_failed_fix_job.rb (hunk 5 above)
Comment (AdamShawBAH) [reply app/jobs/dta_sc_creation_failed_fix_job.rb:85]: thanks yes, didnt want to break loop, will add a next to advance iteration

File: app/jobs/dta_sc_creation_failed_fix_job.rb (hunk 1 above)
Comment (nkutub) [reply app/jobs/dta_sc_creation_failed_fix_job.rb:11]: was this done for testing ? i would prefer it be a constant

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 7]
```diff
@@ -1,36 +1,137 @@
...
-      next unless SupplementalClaim.find_by(
-        decision_review_remanded_id: hlr.id,
-        decision_review_remanded_type: "HigherLevelReview"
-      )
-
-      stuck_job_report_service.append_single_record(hlr.class.name, hlr.id)
-
-      ActiveRecord::Base.transaction do
-        hlr.clear_error!
-      rescue StandardError => error
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:23]: by removing this rescue , one record could break the entire loop, is that what you are intending ?

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 8]
```diff
@@ -1,36 +1,137 @@
...
+    )
+  end
+
+  # Methods for remdiating Appeals with errors
+
+  def process_records_for_appeals_with_errors
+    appeals_with_errors.each do |appeal|
+      if valid_appeal?(appeal)
+        @stuck_job_report_service.append_single_record(appeal.class.name, appeal.id)
+        process_records(appeal)
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:70]: this is another loop that could crash with no rescue and will break the entire loop.

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 9]
```diff
@@ -1,36 +1,137 @@
...
+
+      end_time
+      log_processing_time
+    rescue StandardError => error
+      log_error(error)
+      raise error
+    end
+  end
+
+  def loop_through_and_call_process_records
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:37]: this method name does not match what in the code.

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 10]
```diff
@@ -1,36 +1,137 @@
...
+    begin
+      process_records_for_hlrs_with_errors if hlrs_with_errors.present?
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:39]: these methods should have the term "loop_through_" makes more sense

File: app/jobs/dta_sc_creation_failed_fix_job.rb [hunk 11]
```diff
@@ -1,36 +1,137 @@
...
+      process_records_for_appeals_with_errors if appeals_with_errors.present?
```
Comment (nkutub) [inline app/jobs/dta_sc_creation_failed_fix_job.rb:40]: you have to consecutive loops. what if one errors and the other does not or worse what if one errors on different record as the other. how is that managed ? how will the data look at that point? will the next job run process them correctly ?
